packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart

```diff
@@ -0,0 +1,823 @@
+# deferred imports should be reviewed. Third party plugins and packages may
+# also introduce deferred imports that result in unexpected loading units.
+#
+# This list of loading units are a snapshot of the loading units generated
+# by the latest successful gen_snapshot during the assembly of deferred
+# components opted-in apps.
+loading-units:
+''', flush: false, mode: FileMode.append);
+    final Set<int> usedIds = <int>{};
+    for (final LoadingUnit unit in generatedLoadingUnits) {
+      if (usedIds.contains(unit.id)) {
+        continue;
+      }
+      goldenFile.writeAsStringSync('  - id: ${unit.id}\n', flush: false, mode: FileMode.append);
+      goldenFile.writeAsStringSync('    libraries:\n', flush: false, mode: FileMode.append);
+      for (final String lib in unit.libraries) {
+        goldenFile.writeAsStringSync('      - $lib\n', flush: false, mode: FileMode.append);
+      }
+      usedIds.add(unit.id);
+    }
+  }
+
+  void _clearTempDir() {
+    final Directory dir = env.projectDir.childDirectory('build').childDirectory(kDeferredComponentsTempDirectory);
+    if (dir.existsSync()) {
+      dir.deleteSync(recursive: true);
+    }
+  }
+
+  Future<List<String>> _generateDiff() async {
+    final List<String> diffLines = <String>[];
+    String consumeLog(String line) {
+      diffLines.add(line);
+      // Don't pipe output. We will display the results later.
+      return null;
+    }
+    final List<String> command = <String>[];
+    if (globals.platform.isLinux || globals.platform.isMacOS) {
+      command.add('diff');
+      command.add('-r');
+      command.add('-u');
+      command.add('--unidirectional-new-file');
+      command.add('--exclude=*.diff');
+      command.add('android');
+      command.add('build/$kDeferredComponentsTempDirectory');
+    } else if (globals.platform.isWindows) {
+      // TODO(garyq): find a diff alternative for windows. fc?
+    }
+
+    await globals.processUtils.stream(
+      command,
+      workingDirectory: env.projectDir.path,
+      mapFunction: consumeLog,
+    );
+    if (diffLines.isNotEmpty) {
+      final File diffFile = env.projectDir
+        .childDirectory('build')
+        .childDirectory(kDeferredComponentsTempDirectory)
+        .childFile('setup.diff');
+      if (diffFile.existsSync()) {
+        diffFile.deleteSync();
+      }
+      diffFile.createSync(recursive: true);
+      for (final String line in diffLines) {
+        diffFile.writeAsStringSync('$line\n', mode: FileMode.append, flush: true);
+      }
+    }
+    return diffLines;
+  }
+
+  static const String _thickDivider = '=================================================================================';
+  static const String _thinDivider = '---------------------------------------------------------------------------------';
+
+  // TODO(garyq): Gradle and re-entrant Flutter currently eats up any color and
+  // text formatting/highlighting. Figure out how to preserve this info.
```

Comment:
If you use printError that will end up with error/red text. I would not spend too much time specifically formatting the errors